lib/KettleMiddleware-Multer.js

```diff
@@ -0,0 +1,149 @@
+/*!
+Kettle wrapping for Multer Express Middleware
+
+Copyright 2017-2018 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/fluid-project/kettle/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+// Wraps the standard Express Multer middleware, for handling
+// multipart/form-data (primarily for file uploads)
+
+var fluid = require("infusion"),
+    kettle = fluid.registerNamespace("kettle");
+
+fluid.require("multer", require, "kettle.npm.multer");
+
+fluid.defaults("kettle.middleware.multer", {
+    gradeNames: "kettle.plainAsyncMiddleware",
+    // See https://github.com/expressjs/multer#multeropts; because
+    // Multer expects to receive functions for some of its options,
+    // some of these are implemented using invokers that an implementer should
+    // override to change the basic configuration
+    middlewareOptions: {
+        // These are the default limits multer uses
+        // See https://github.com/expressjs/multer#limits
+        // Recommended that these should be configured on a per-handler basis
+        // to prevent possible DDOS attacks via form submission, or other
+        // badness
+        limits: {
+            fieldNameSize: 100,
+            fieldSize: "1MB",
+            fields: Infinity,
+            fileSize: Infinity,
+            files: Infinity,
+            parts: Infinity,
+            headerPairs: 2000
+        }
+    },
+    middleware: "@expand:kettle.middleware.multer.createMiddleware({that}, {that}.options.middlewareOptions, {that}.getStorage, {that}.getMiddlewareForFileStrategy, {that}.getFileFilter)",
+    invokers: {
+        // Change to getDiskStorage for diskStorage
+        "getStorage": {
+            "func": "{that}.getMemoryStorage"
+        },
+        //
+        "getMiddlewareForFileStrategy": {
+            "funcName": "kettle.middleware.multer.getMiddlewareForFileStrategy",
+            "args": ["{that}", "single", "file"]
+        },
+        // Override this function for different behaviour in the file filter
+        // function
+        "getFileFilter": {
```

Comment:
These two are a bit peculiar and indirect - the invoker can simply be fileFilter: {
    func: "kettle.npm.multer.allowAll"
}
etc.

Reply:
Digging in further, this actually didn't even work at all - Multer doesn't export its own allowAll function (https://github.com/expressjs/multer/blob/80ee2f52432cc0c81c93b03c6b0b448af1f626e5/index.js#L7), but the issue was invisible because a "null" function for the file filter wasn't registered by Multer as a problem, as it defaulted to its internal allowAll (the one it doesn't export): https://github.com/expressjs/multer/blob/80ee2f52432cc0c81c93b03c6b0b448af1f626e5/index.js#L22

Middleware integration is fun! I'll clean this up and try to be less indirect and peculiar.

Reply:
I think I've accomplished this refactoring, which has reduced the size of the file and the number of functions considerably. Thanks for the suggestions! My only uncertainty is described at https://github.com/waharnum/kettle/blob/KETTLE-66/lib/KettleMiddleware-Multer.js#L46-L58 - is this the best way to accomplish what's needed? (passing in a keyed configuration object to a third-party library's function)